### DISCUSSION TITLE
Creation of backup from CLI

### DISCUSSION THREAD

**Discussion by Daniel-I-Am** [dani-garcia/vaultwarden#3409] [Ideas]
I stumbled upon https://github.com/dani-garcia/vaultwarden/issues/246 and want to request that a system for creating backups from the Vaultwarden CLI is added to that list.

## Some context
Right now I am running the Vaultwarden Docker image (`vaultwarden/server:1.28.0`). This system has been working great for me, except for one part: backups.

I was unable to find any neat way to automate the backups of my Vaultwarden instance. And for me solid, automated, testable backups of a system like my password vault is a must. So I bodged together a script that works for me that is run from a user that has exec permissions to the Docker container.

This script picks up the `ADMIN_TOKEN` environment variable to authenticate to the admin web interface. As using the admin interface is the only way that I've found to make a backup from within the container (as SQLite tools are not installed in the container to make a dump following the usual instructions found [here](https://github.com/dani-garcia/vaultwarden/wiki/General-%28not-docker%29)). This is awful and I've been looking to retire it for some time (basically since the day I wrote it).

When I updated to the latest version of Vaultwarden recently, I saw the headsup about retiring the 'ADMIN_TOKEN' environment variable. Currently this script - or the lack of a way to do automated backups in general - prohibits me from moving away from `ADMIN_TOKEN` and onto the new PHC string format for increased security.

As I haven't found any way to automate backups without knowing the admin token and going through the admin web interface, I have opened this request for a better solution to be added to the feature requests ticket.

---

My current hacky backup script snippet (don't use this, if you can avoid it).
```
sudo docker exec -i VaultwardenServer bash <<'EOF_DOCKER'
set -ex

urlencodepipe() {
  local LANG=C; local c; while IFS= read -r c; do
    case $c in [a-zA-Z0-9.~_-]) printf "$c"; continue ;; esac
    printf "$c" | od -An -tx1 | tr ' ' % | tr -d '\n'
  done <<EOS
$(fold -w1)
EOS
  echo
}

COOKIE_STORE="/tmp/cookie-store-$RANDOM"
curl --fail -o /dev/null -X POST --cookie-jar "$COOKIE_STORE" --cookie "$COOKIE_STORE" 'http://localhost/admin' --data-raw "token=$(echo "$ADMIN_TOKEN" | urlencodepipe)"
curl --fail -X POST --cookie-jar "$COOKIE_STORE" --cookie "$COOKIE_STORE" 'http://localhost/admin/config/backup_db'
rm -f "$COOKIE_STORE"
EOF_DOCKER

BACKUP_FILE="$(sudo find /srv/docker-volumes/vaultwarden/data/ -name 'db_*.sqlite3' -exec ls -t {} \+ | head -1)"
```

## My request
Can we get some option added to the Vaultwarden CLI, so we can create a backup (the same one that is made by clicking 'Backup Database' in the admin web interface).

My proposal, a command that would somewhat look like this, to facilitate the creation of a backup of the SQLite (or whatever backend you use) database.
```
/vaultwarden backup --output-location /mnt/backups/backup-2023-04-02.sqlite3
```

**Comment by stefan0xC** [comment]
Have you looked at [the existing third party backup examples from the wiki](https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#examples)?

**Comment by BlackDex** [comment]
This is on my personal list somewhere to add.
Also via a usr kill signal maybe.

But not a priority for my self.
It will however only be for sqlite. All other databases need to manage backups them selfs.

**Reply by Daniel-I-Am** [reply]
All other database backends are stored externally and have their own backup solutions. That's why my question arose for SQLite, as it is the only database which is not running seperate from the Vaultwarden container. Therefor it is more difficult to back up.

Great to hear that it's on your todo list. But in the meantime, I'll take a look at third party options as someone else mentioned here.

**Reply by BlackDex** [reply]
Also see https://github.com/dani-garcia/vaultwarden/wiki/Backing-up-your-vault#sqlite-database-files
